core/src/main/scala/org/apache/spark/executor/Executor.scala

```diff
@@ -130,6 +130,14 @@ private[spark] class Executor(
   private val urlClassLoader = createClassLoader()
   private val replClassLoader = addReplClassLoaderIfNeeded(urlClassLoader)
 
+  // Load plugins in the current thread, they are expected to not block.
+  // Heavy computation in plugin initialization should be done async.
+  Thread.currentThread().setContextClassLoader(replClassLoader)
+  conf.get(EXECUTOR_PLUGINS).foreach { classes =>
```

Comment:
This becomes more readable if the config has a `Nil` default value instead of being optional.

Reply:
@vanzin If I'm understanding correctly, in the ConfigBuilder I am to change `.createOptional` to `.createWithDefault(Nil)`. This gives the error:

`[error] /Users/nsheth/personal_fork/spark/core/src/main/scala/org/apache/spark/executor/Executor.scala:136: type mismatch;
[error]  found   : String
[error]  required: Seq[String]
[error]     Utils.loadExtensions(classOf[ExecutorPlugin], classes, conf)`

Granted I have no idea why it's able to cast a Nil to a string for this config.

Reply:
@mridulm We've had some back and forth on that, open to changing it if people agree it should be one way over another. Just checking, have you looked at the earlier conversation at https://github.com/apache/spark/pull/22192#discussion_r212746543? There's also one on the old PR for the same topic.

As for moving plugins for after executors have initialized, I see no problem with that. I don't think it should make a huge difference provided either we do change to separate thread, or we keep it in the same thread but writers respect not making the initialization blocking and computation heavy. I'll see if there's a more natural place to move it

Reply:
It's not just changing the config. You'd need to change the call too. e.g.

    Utils.loadExtensions(classOf[ExecutorPlugin], conf.get(EXECUTOR_PLUGINS), conf)

Reply:
Ah, whoops, misunderstood the value of `conf.get`. That works just fine